src/ucp/core/ucp_mm.c

```diff
@@ -400,99 +425,152 @@ ucp_memh_get_slow(ucp_context_h context, void *address, size_t length,
 
     UCP_THREAD_CS_ENTER(&context->mt_lock);
     if (context->rcache == NULL) {
-        memh = ucs_calloc(1, sizeof(*memh) +
-                          (sizeof(uct_mem_h) * context->num_mds), "ucp_rcache");
-        if (memh == NULL) {
-            status = UCS_ERR_NO_MEMORY;
-            goto out;
-        }
-
-        memh->super.super.start = (uintptr_t)reg_address;
-        memh->super.super.end   = (uintptr_t)reg_address + reg_length;
-        memh->alloc_md_index    = UCP_NULL_RESOURCE;
-        memh->alloc_method      = UCT_ALLOC_METHOD_LAST;
-        memh->map_count         = 0;
+        status = ucp_memh_create_no_rcache(context, reg_address, reg_length,
+                                           mem_type, &memh);
+        memh->parent = memh;
     } else {
-        status = ucs_rcache_get_unsafe(context->rcache, reg_address, reg_length,
-                                       PROT_READ | PROT_WRITE, NULL, &rregion);
-        if (status != UCS_OK) {
-            goto out;
-        }
+        status = ucp_memh_create_rcache(context, reg_address, reg_length,
+                                        mem_type, &memh);
+    }
 
-        memh        = ucs_derived_of(rregion, ucp_mem_t);
-        reg_address = (void*)rregion->super.start;
-        reg_length  = rregion->super.end - rregion->super.start;
+    if (status != UCS_OK) {
+        goto out;
     }
 
-    memh->mem_type = mem_type;
-    status         = ucp_memh_register(context, memh,
-                                       ~memh->md_map & reg_md_map,
-                                       reg_address, reg_length,
-                                       uct_flags);
-    if (status == UCS_OK) {
-        ucs_trace("memh %p: registered address %p/%p length %zu/%zu md_map %"
-                  PRIx64 "/%" PRIx64, memh, reg_address, ucp_memh_address(memh),
-                  reg_length, ucp_memh_length(memh), reg_md_map, memh->md_map);
-        *memh_p = memh;
+    /* Reinitialize address and length, because they could be greater in case
+     * a region which includes the searched region was found in the rcache */
+    reg_address = ucp_memh_address(memh);
+    reg_length  = ucp_memh_length(memh);
+
+    ucs_assert(memh->mem_type == mem_type);
+
+    status = ucp_memh_register(context, memh, ~memh->md_map & reg_md_map,
+                               reg_address, reg_length, uct_flags);
+    if (status != UCS_OK) {
+        goto err_free_memh;
     }
 
+    ucs_trace("memh %p: registered address %p/%p length %zu/%zu md_map %"
+              PRIx64 "/%" PRIx64, memh, reg_address, ucp_memh_address(memh),
+              reg_length, ucp_memh_length(memh), reg_md_map, memh->md_map);
+    *memh_p = memh;
+
 out:
     UCP_THREAD_CS_EXIT(&context->mt_lock);
     return status;
+
+err_free_memh:
+    if (context->rcache == NULL) {
+        ucs_free(memh);
```

Comment:
maybe add helper function ucp_memh_free that asserts md_map==0 and then does ucs_free(memh)
seems it would be useful to detect leaked registrations especially in error flow